okhttp/src/main/java/okhttp3/internal/ws/WebSocketExtensions.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.io.IOException
+import okhttp3.Response
+import okhttp3.internal.delimiterOffset
+import okhttp3.internal.trimSubstring
+
+/**
+ * Models the contents of a `Sec-WebSocket-Extensions` response header. OkHttp honors one extension
+ * `permessage-deflate` and two of its parameters, `client_no_context_takeover` and
```

Comment:
Again, for future me. `12` here is just any value 8<=X<15.

window_bits | sender=15 | sender=12
-------------------- | ------- | -------- 
receiver=15  | ok | ok
receiver=12 | failure | ok

For client-server stream, okhttp is a **sender** and can use any window_size, as long as it is supported by server.

For server-client stream, okhttp is a **receiver** and can work with any `server_max_window_bits` received from server. Server won't send `client_max_window_bits` without prior request from client.

So, 
1. There is no reason for okhttp to send `server_max_window_bits` or `client_max_window_bits` to server, as okhttp supports 15 (and therefore, any value when okhttp is reading).
2. okhttp MAY ignore `server_max_window_bits` from server, as it defines the sliding window for server-client stream, and okhttp is fine with it being smaller than 15. However, we MUST NOT fail if we surprisingly receive such parameter from server. [7.1.2.1, last paragraph](https://tools.ietf.org/html/rfc7692#section-7.1.2.1)
3. okhttp MUST fail on `client_max_window_bits` from server, as server MUST NOT send `client_max_window_bits` unless received it from client in request (and we don't send it).

Reply:
@swankjesse , 

looks like support for 2 must be added to avoid failing in cases where we can actually operate normally. 

The logic is fine for 3, but I'd update the reasoning for `client_max_window_bits` in the doc below to something like: "client_max_window_bits is unknown because server MUST NOT return it unless requested by client, and we don't send it".

Reply:
Though, for 2, I'm not sure how to test it live, eg where we can find a server which will have configurable LZ77 sliding window.

Reply:
Oooh good catch. We should accept and ignore `server_max_window_bits=8` through `server_max_window_bits=15`. We can't benefit from allocating a smaller buffer due to Java API limitations.

Reply:
Updated to support parsing both the client_max_window_bits and server_max_window_bits. It’s simpler to just parse everything and to do validation work separately.